mozilla/fx-private-relay-add-on
# Feature: Re-use/autofill previously generated alias

### Summary: 

This feature allows users to re-use aliases via context menu (alt-click)

### Major changes: 
- Moved all context menu code to separate background script 
- For non-premium accounts, the "Generate New Alias" will now stay visible but disabled after user has created maximum amount of aliases
- Added new top-level menu item: "Manage all aliases" that takes user to Relay account page 

#### TODO: 
- [x] Migrate strings to L10N 
- [x] Add logic for site-specific aliases
- [x] (Optional) Work with API to query aliases each interaction item rather than relay on local storage data
- [x] Add "Logged in" check during init
- [x] Only call `relayContextMenus.utils.getAliases` once during init. 
- [x] BUG: If menu is called twice, "Use on this site…" breaks. 

## Testing Steps:

- Start with a non-premium account with ZERO aliases created 
- Navigate to login form on a website (Ex: [Spotify](https://accounts.spotify.com/en/login), [Imgur](https://imgur.com/signin))
- Right/alt-click on Email field
- **Expected Menu:** Three items: Generate Alias, Mange All Aliases, Get Unlimited Aliases
- Right/alt-click: Generate alias
- **Expected:** New alias is created and populated into field. (You can validate this on the dashboard too) 
- Clear field, refresh page. 
- Right/alt-click on Email field again.
- **Expected menu:** should now include the Generate Alias, Use existing alias from this site…, Mange All Aliases, Get Unlimited Aliases
- Go to dashboard.
- Create a maximum amount of aliases, give some labels
- Go to a different site than the first 
- Right/alt-click on Email field
- **Expected menu** Generate Alias, Recent Alias, Mange All Aliases, Get Unlimited Aliases
  - Generate Alias should be greyed out
- Upgrade user account premium
- Right/alt-click on Email field
- **Expected menu** Generate Alias, Recent Alias, Mange All Aliases
  - Generate Alias should be back to normal

## Screenshots

_Note – The yellow background is OS-specific theme highlight and not part of this PR/code base_  

Non-premium Account: Generate New Alias is enabled
<img width="458" alt="image" src="https://user-images.githubusercontent.com/2692333/144298558-a6a0659c-25aa-4a69-99ae-5fd16acb1f22.png">

Non-premium Account: Generate New Alias is disabled as max alias count is reached 
<img width="444" alt="image" src="https://user-images.githubusercontent.com/2692333/144298662-15ec1ce8-eaba-4e6a-b4e9-77f9771294e2.png">

Premium Account: Generate New Alias is enabled (and upgrade button is removed)
<img width="439" alt="image" src="https://user-images.githubusercontent.com/2692333/144299028-d735ee52-7c45-43d2-a574-f45e22d44f45.png">

Premium Account: Reuse alias for site where no previous aliases were generated for: 
<img width="525" alt="image" src="https://user-images.githubusercontent.com/2692333/144960946-f3ad6a40-9094-4049-bb7e-0265f4b67be9.png">

Premium Account: Reuse alias for site where some previous aliases were generated for: 
<img width="615" alt="image" src="https://user-images.githubusercontent.com/2692333/144960972-02eaa58e-c22e-46b5-bb7f-1c4401b1a31c.png">
